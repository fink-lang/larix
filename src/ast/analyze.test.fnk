{describe, it, expect, to_match_snapshot} = import '@fink/jest/test.fnk'


{parse} = import '../parser.fnk'

{add_parents, add_binding_refs, add_duplicate_bindings, get_expr_id, get_scope} = import './analyze.fnk'
{walk} = import './traverse.fnk'



collect_exprs = fold [expr], , , ctx={}: [expr, , ctx]


collect_bindings = fn parsed:
  [, {bindings}] = collect_exprs parsed
  [...bindings | map expr: get_expr_id expr]



collect_refs = fn parsed:
  [, {refs}] = collect_exprs parsed

  seq:
    ...refs | map [expr_id, exprs]:
      ... exprs | map expr:
        '${get_expr_id expr} -> ${expr_id}'



collect_scopes = fn parsed:
  [, ctx] = collect_exprs parsed
  seq:
    ... ctx.bindings | map expr:
      '${get_expr_id expr}  ${get_expr_id get_scope expr, ctx}'



collect_duplicate_bindings = fn parsed:
  [, ctx] = collect_exprs parsed
  seq:
    ... ctx.rebindings | map expr:
      '${get_expr_id expr}  ${get_expr_id get_scope expr, ctx}'




analyze = fn source, collector:
  pipe source:
    parse ?, 'test.fnk'
    walk [?]
    add_parents
    add_binding_refs
    add_duplicate_bindings
    collector




source = '
  foo = 1234

  bar = fn spam, ham:
    ni = spam.ni + ham"\${foo}"
    ni + undef

  {ni, na: {nu}} = {foo}
  {(ni): nini=nu} = {foo: bar, (ni): nu}

  match bar foo:
    nu:
      ni = foo
      ni

  <div />

  Spam = fn {children}:
    <>{children}</>

  <Spam ni shrub=nu spam={ham + 1}>
  </Spam>

  try:
    foo

  [ni] = [1234]
'


describe analyze, fn:

  it 'gets bindings', fn:
    expect
      analyze source, collect_bindings
      to_match_snapshot


  it 'gets duplicate bindings', fn:
    expect
      analyze source, collect_duplicate_bindings
      to_match_snapshot


  it 'gets refs', fn:
    expect
      analyze source, collect_refs
      to_match_snapshot


  it 'gets scopes', fn:
    expect
      analyze source, collect_scopes
      to_match_snapshot

