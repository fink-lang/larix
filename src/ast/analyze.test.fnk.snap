// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze gets bindings 1`] = `
Array [
  "ident:foo@1:0-1:3",
  "ident:bar@3:0-3:3",
  "ident:spam@3:9-3:13",
  "ident:ham@3:19-3:22",
  "ident:ni@4:2-4:4",
  "ident:ni@7:1-7:3",
  "ident:nu@7:10-7:12",
  "ident:nini@8:7-8:11",
  "ident:ni@12:4-12:6",
  "ident:Spam@18:0-18:4",
  "ident:children@18:11-18:19",
  "ident:ni@27:1-27:3",
  "ident:na@28:0-28:2",
  "ident:foo@29:3-29:6",
  "ident:ham@31:10-31:13",
]
`;

exports[`analyze gets duplicate bindings 1`] = `
Array [
  "ident:ni@27:1-27:3  module@1:0-33:0",
]
`;

exports[`analyze gets refs 1`] = `
Array [
  "ident:foo@4:26-4:29 -> ident:foo@1:0-1:3",
  "ident:foo@7:18-7:21 -> ident:foo@1:0-1:3",
  "ident:foo@10:10-10:13 -> ident:foo@1:0-1:3",
  "ident:foo@12:9-12:12 -> ident:foo@1:0-1:3",
  "ident:foo@25:2-25:5 -> ident:foo@1:0-1:3",
  "ident:bar@8:24-8:27 -> ident:bar@3:0-3:3",
  "ident:bar@10:6-10:9 -> ident:bar@3:0-3:3",
  "ident:spam@4:7-4:11 -> ident:spam@3:9-3:13",
  "ident:ham@4:20-4:23 -> ident:ham@3:19-3:22",
  "ident:ni@5:2-5:4 -> ident:ni@4:2-4:4",
  "ident:undef@5:7-5:12 -> unbound",
  "jsx:elem:div@16:0-16:7 -> unbound",
  "ident:ham@21:24-21:27 -> unbound",
  "ident:ni@8:2-8:4 -> ident:ni@7:1-7:3",
  "ident:ni@8:30-8:32 -> ident:ni@7:1-7:3",
  "jsx:attr:ni@21:6-21:8 -> ident:ni@7:1-7:3",
  "ident:nu@8:12-8:14 -> ident:nu@7:10-7:12",
  "ident:nu@8:35-8:37 -> ident:nu@7:10-7:12",
  "ident:nu@11:2-11:4 -> ident:nu@7:10-7:12",
  "ident:nu@21:15-21:17 -> ident:nu@7:10-7:12",
  "ident:ni@13:8-13:10 -> ident:ni@12:4-12:6",
  "jsx:elem:Spam@21:0-22:7 -> ident:Spam@18:0-18:4",
  "ident:children@19:5-19:13 -> ident:children@18:11-18:19",
  "ident:ni@28:9-28:11 -> ident:ni@27:1-27:3",
  "ident:foo@29:9-29:12 -> ident:foo@29:3-29:6",
  "ident:ham@32:2-32:5 -> ident:ham@31:10-31:13",
]
`;

exports[`analyze gets scopes 1`] = `
Array [
  "ident:foo@1:0-1:3  module@1:0-33:0",
  "ident:bar@3:0-3:3  module@1:0-33:0",
  "ident:spam@3:9-3:13  block:fn@3:6-5:12",
  "ident:ham@3:19-3:22  block:fn@3:6-5:12",
  "ident:ni@4:2-4:4  block:fn@3:6-5:12",
  "ident:ni@7:1-7:3  module@1:0-33:0",
  "ident:nu@7:10-7:12  module@1:0-33:0",
  "ident:nini@8:7-8:11  module@1:0-33:0",
  "ident:ni@12:4-12:6  match:expr@11:2-13:11",
  "ident:Spam@18:0-18:4  module@1:0-33:0",
  "ident:children@18:11-18:19  block:fn@18:7-19:17",
  "ident:ni@27:1-27:3  module@1:0-33:0",
  "ident:na@28:0-28:2  module@1:0-33:0",
  "ident:foo@29:3-29:6  block:fn@29:0-29:13",
  "ident:ham@31:10-31:13  block:map@31:0-33:0",
]
`;
